OsmAnd/src/net/osmand/plus/server/ServerFragment.java

```
@@ -0,0 +1,117 @@
+				}
+			}
+		});
+		view.findViewById(R.id.server_stop_button).setOnClickListener(new View.OnClickListener() {
+			@Override
+			public void onClick(View view) {
+				updateTextView(getString(R.string.click_button_to_start_server));
+				deInitServer();
+			}
+		});
+		return view;
+	}
+
+	@Override
+	public void onDestroy() {
+		deInitServer();
+		super.onDestroy();
+	}
+
+	public static void enableStrictMode() {
+		StrictMode.setThreadPolicy(
+				new StrictMode.ThreadPolicy.Builder()
+						.detectDiskReads()
+						.detectDiskWrites()
+						.detectNetwork()
+						.penaltyLog()
+						.build());
+		StrictMode.setVmPolicy(
+				new StrictMode.VmPolicy.Builder()
+						.detectLeakedSqlLiteObjects()
+						.penaltyLog()
+						.build());
+	}
+
+	private void updateTextView(String text) {
+		((TextView) view.findViewById(R.id.server_status_textview)).setText(text);
+	}
+
+	private void initServer() {
+		final int THREAD_ID = 10000;
+		TrafficStats.setThreadStatsTag(THREAD_ID);
+		OsmAndHttpServer.HOSTNAME = getDeviceAddress();
+		try {
+			server = new OsmAndHttpServer();
+			server.setApplication((OsmandApplication) getMyApplication());
+			initialized = true;
+			updateTextView("Server started at: http://" + getDeviceAddress() + ":" + OsmAndHttpServer.PORT);
+		} catch (IOException e) {
+			Toast.makeText(requireContext(),
+					e.getLocalizedMessage(),
+					Toast.LENGTH_SHORT).show();
+			e.printStackTrace();
+		}
+	}
+
+	private String getDeviceAddress() {
+		WifiManager wm = (WifiManager) requireContext().getSystemService(WIFI_SERVICE);
+		String ip = Formatter.formatIpAddress(wm.getConnectionInfo().getIpAddress());
+		return ip != null ? ip : "0.0.0.0";
+	}
+
+	private void deInitServer() {
+		if (server != null) {
+			server.closeAllConnections();
+			server.stop();
+		}
+		initialized = false;
+		if (getActivity() != null) {
+			try {
+				getActivity().getSupportFragmentManager().beginTransaction().remove(this).commit();
+			} catch (Exception e) {
```

??? swallowing exception ???